### PR TITLE
S3 processor

### PR DESCRIPTION
**Pull request by Justin-Holmes**
Related story: https://github.com/ello/kinesis-stream-reader/pull/17

### CODE REVIEW

File: lib/ello/kinesis_consumer/s3_processor.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+module Ello
+  module KinesisConsumer
+    class S3Processor < BaseProcessor
+
+      def run!
+        @stream_reader.run! do |record, opts|
+          @logger.debug "#{opts[:schema_name]}: #{record}"
+          obj = s3_bucket.object(opts[:sequence_number])
+          obj.put(body: opts[:raw_data])
```
Comment (alanpeabody) [inline lib/ello/kinesis_consumer/s3_processor.rb]: Do we care about which shard they are from - or does sequence number cover that?

Comment (Justin-Holmes) [inline lib/ello/kinesis_consumer/s3_processor.rb]: The sequence number determines which shard it goes too, so I don't think so. We store the filename in S3 as the sequence number, so in the event we reprocess it should go to the same shard no matter what.

Comment (alanpeabody) [inline lib/ello/kinesis_consumer/s3_processor.rb]: @Justin-Holmes my understanding is that sequence numbers are unique within a shard, but not a stream. I think the partition key is what determines which shard.

http://docs.aws.amazon.com/streams/latest/dev/key-concepts.html

I think that means our key needs to include the shard, maybe `"#{opts[:shard_id]}/#{opts[:sequence_number]}"` or similar?

Comment (Justin-Holmes) [inline lib/ello/kinesis_consumer/s3_processor.rb]: @alanpeabody Good catch, I obviously got partition key and sequence number confused. I'll make the change 👍

**Comment by Justin-Holmes** [comment]
@alanpeabody @jayzes This worked like a charm on staging. Once we merge in https://github.com/ello/kinesis-stream-reader/pull/17 I can update the Gemfile and then we should be good to go.

I ended up not adding encryption to our S3 uploads. If you guys think it's worth adding that extra layer of security then I can add it. It doesn't seem too involved, but we will need to be sure to store the generated key in 1Password: http://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Encryption.html

File: lib/ello/kinesis_consumer/s3_processor.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+          @logger.debug "#{opts[:schema_name]}: #{record}"
+          obj = s3_bucket.object("#{opts[:shard_id]}/#{opts[:sequence_number]}")
+          obj.put(body: opts[:raw_data])
+        end
+      end
+
+      private
+
+      def s3_bucket
+        @s3_bucket ||= Aws::S3::Resource.new.bucket(ENV['KINESIS_STREAM_NAME'])
```
Comment (jayzes) [inline lib/ello/kinesis_consumer/s3_processor.rb]: Should we make the bucket name configurable to be different than the stream name?

File: lib/ello/kinesis_consumer/s3_processor.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'aws-sdk'
+
+module Ello
+  module KinesisConsumer
+    class S3Processor < BaseProcessor
+
+      def run!
+        @stream_reader.run! do |record, opts|
+          @logger.debug "#{opts[:schema_name]}: #{record}"
+          obj = s3_bucket.object("#{opts[:shard_id]}/#{opts[:sequence_number]}")
```
Comment (jayzes) [inline lib/ello/kinesis_consumer/s3_processor.rb]: Worth including the stream name in here as well so we could point multiple streams at the same bucket and not have them get tangled?

File: lib/ello/kinesis_consumer/s3_processor.rb (hunk 2 above)
Comment (Justin-Holmes) [inline lib/ello/kinesis_consumer/s3_processor.rb]: 👍

**Comment by Justin-Holmes** [comment]
@jayzes Made those changes, LMK what you think!

**Comment by jayzes** [comment]
Looks good!

**Pull request merged**